### ISSUE TITLE
Allow to pass if the branch does not exist

### ISSUE THREAD

**Issue by AlexSkrypnyk** [drevops/git-artifact#92]
## Problem

In CI/CD workflows, deployment may fail when a PR is merged and the source branch is deleted while the deployment job is still running. This is a legitimate concurrency issue that affects teams using automated PR merging with branch deletion.

### How GitHub Actions Works:
1. When a workflow starts, GitHub Actions checks out code by **commit SHA**, not by branch name
2. This results in a **detached HEAD** state: `(HEAD detached at abc1234)`
3. Even if the branch is deleted on the remote, the workflow continues running on that commit
4. Branch deletion does NOT stop running workflows - they continue on the commit SHA

### Why git-artifact Currently Fails:

When `getOriginalBranch()` is called in detached HEAD state:
1. Detects detached HEAD from `getCurrentBranchName()` output
2. Tries to extract the branch name using regex pattern matching
3. Expected patterns that work:
   - `(HEAD detached at feature/my-branch)` ✅ - can extract branch name
   - `(HEAD detached at v1.0.0)` ✅ - can extract tag name
4. Pattern that fails:
   - `(HEAD detached at abc1234)` ❌ - only commit hash, no branch info
5. Throws exception: `"Unable to determine a detachment source"`
6. Deployment fails with exit code 1

### Common Scenario:
```
1. PR opened on feature/my-branch
2. CI starts: build → test → deploy jobs
3. Test job passes ✅
4. Bot/automation merges PR and deletes feature/my-branch (30-60 seconds)
5. Deploy job starts running git-artifact
6. Git-artifact runs in detached HEAD (commit abc1234)
7. Cannot determine original branch name
8. Deployment fails ❌ (exit code 1)
9. CI pipeline marked as failed
```

**Note:** This happens regardless of how the branch is checked out. Whether it's detached HEAD or a full branch checkout, if the source branch no longer exists, we cannot determine it.

## Solution

Add `--fail-on-missing-branch` flag. **Change default behavior** to gracefully skip deployment when branch is missing (backwards incompatible change).

### Default Behavior (WITHOUT flag) - NEW:
```bash
git-artifact remote-url
```
- Branch not found → **Skip deployment**
- Exit code: **0** (success)
- Message: `"Source branch not found. Deployment skipped."`
- Shows commit hash for reference
- CI pipeline continues successfully
- No push to remote repository

### With `--fail-on-missing-branch` Flag - OLD behavior:
```bash
git-artifact --fail-on-missing-branch remote-url
```
- Branch not found → **Fail deployment**
- Exit code: **1** (failure)
- Message: `"Unable to determine source branch. Deployment failed."`
- CI pipeline marked as failed
- No push to remote repository

### Rationale:
- Most teams want CI to succeed when branch is deleted (PR was merged successfully)
- Failing the deployment after successful merge is confusing
- Teams that need strict branch tracking can opt-in with the flag
- Consistent behavior regardless of checkout method (detached HEAD or branch)

### 1. Custom Exception Hierarchy
Create exceptions in `src/Exception/`:
- `GitArtifactException` - Base exception
- `GitException` - Git-related errors
- `BranchNotFoundException` - Cannot determine branch (stores commit hash)

### 2. Update `getOriginalBranch()`
- Keep existing logic unchanged
- When branch cannot be determined, get current commit hash
- Throw `BranchNotFoundException` with commit hash included

### 3. Handle Exception in `ArtifactCommand`
- Catch `BranchNotFoundException` in `resolveOptions()`
- If `--fail-on-missing-branch`: re-throw as RuntimeException (fail)
- If NOT set (default): log message, show output, exit(0) (skip)

### 4. Update Tests
- Test default behavior: branch missing → skip deployment → exit 0
- Test with flag: branch missing → fail deployment → exit 1
- Test normal case: branch exists → deploy → exit 0

## Breaking Change

This is a **breaking change**. The default behavior changes from:
- **Before**: Deployment fails (exit 1) when branch missing
- **After**: Deployment skips gracefully (exit 0) when branch missing

Teams that need the old behavior must add `--fail-on-missing-branch` flag.

## Out of Scope

Cleanup in detached HEAD state has a separate issue where it doesn't properly handle checking out commit hashes. This will be addressed separately and is NOT part of this issue.
